crates/noirc_evaluator/src/ssa_refactor/opt/unrolling.rs

```diff
@@ -0,0 +1,575 @@
+        //     v6 = add v2, Field 1
+        //     jmp b4(v6)
+        //   b6(): // end of inner loop
+        //     v7 = add v0, Field 1
+        //     jmp b1(v7)
+        //   b3(): // end of outer loop
+        //     return Field 0
+        // }
+        let main_id = Id::test_new(0);
+
+        // Compiling main
+        let mut builder = FunctionBuilder::new("main".into(), main_id);
+
+        let b1 = builder.insert_block();
+        let b2 = builder.insert_block();
+        let b3 = builder.insert_block();
+        let b4 = builder.insert_block();
+        let b5 = builder.insert_block();
+        let b6 = builder.insert_block();
+
+        let v0 = builder.add_block_parameter(b1, Type::field());
+        let v2 = builder.add_block_parameter(b4, Type::field());
+
+        let zero = builder.field_constant(0u128);
+        let one = builder.field_constant(1u128);
+        let three = builder.field_constant(3u128);
+        let four = builder.field_constant(4u128);
+        let ten = builder.field_constant(10u128);
+
+        builder.terminate_with_jmp(b1, vec![zero]);
+
+        // b1
+        builder.switch_to_block(b1);
+        let v1 = builder.insert_binary(v0, BinaryOp::Lt, three);
+        builder.terminate_with_jmpif(v1, b2, b3);
+
+        // b2
+        builder.switch_to_block(b2);
+        builder.terminate_with_jmp(b4, vec![zero]);
+
+        // b3
+        builder.switch_to_block(b3);
+        builder.terminate_with_return(vec![zero]);
+
+        // b4
+        builder.switch_to_block(b4);
+        let v3 = builder.insert_binary(v2, BinaryOp::Lt, four);
+        builder.terminate_with_jmpif(v3, b5, b6);
+
+        // b5
+        builder.switch_to_block(b5);
+        let v4 = builder.insert_binary(v0, BinaryOp::Add, v2);
+        let v5 = builder.insert_binary(ten, BinaryOp::Lt, v4);
+        builder.insert_constrain(v5);
+        let v6 = builder.insert_binary(v2, BinaryOp::Add, one);
+        builder.terminate_with_jmp(b4, vec![v6]);
+
+        // b6
+        builder.switch_to_block(b6);
+        let v7 = builder.insert_binary(v0, BinaryOp::Add, one);
+        builder.terminate_with_jmp(b1, vec![v7]);
+
+        // basic_blocks_iter iterates over unreachable blocks as well, so we must filter those out.
+        let count_reachable_blocks = |ssa: &Ssa| {
+            let function = ssa.main();
+            let dom_tree = DominatorTree::with_function(function);
+            function
+                .dfg
+                .basic_blocks_iter()
+                .filter(|(block, _)| dom_tree.is_reachable(*block))
+                .count()
+        };
+
+        let ssa = builder.finish();
+        assert_eq!(count_reachable_blocks(&ssa), 7);
+
+        // The final block count is not 1 because the block creates some unnecessary jmps.
+        // If a simplify cfg pass is ran afterward, the expected block count will be 1.
+        let ssa = ssa.unroll_loops();
+        assert_eq!(count_reachable_blocks(&ssa), 5);
+    }
```

Comment:
If I add a `println!("{}", ssa.main());` here, I see a pretty broken looking result:
```
fn main f0 {
  b0():
    jmp b1(Field 0)
  b1(v0: Field):
    v7 = lt v0, Field 3
    jmpif v7 then: b2, else: b3
  b2():
    jmp b8()
  b8():
    (no terminator instruction)
  b3():
    return Field 0
}
```

Reply:
Good catch, that output is definitely wrong. I think something must have gotten messed up in c270aae and this bad output just happened to have the same amount of reachable blocks as the good output so it wasn't caught.

Reply:
The bug should be fixed now. The issue was, with the addition of `fresh_block` when unrolling headers, we never set `self.insert_block` back to `unroll_into` after the fresh_block was emptied and thrown away. So the rest of the loop body was being inserted into an unreachable block.